tests/misc.js

```diff
@@ -478,3 +478,266 @@ describe('literal JSON', () => {
+
+    assert.equal(expansionMapCalled, true);
+  });
+
+  it('should be called on nested un-mapped term', async () => {
+    const docWithUnMappedTerm = {
+      '@context': {
+        'definedTerm': 'https://example.com#definedTerm'
+      },
+      definedTerm: {
+        testUndefined: "is undefined"
+      }
+    };
+
+    let expansionMapCalled = false;
+    const expansionMap = info => {
+      if(info.unmappedProperty === 'testUndefined') {
+        expansionMapCalled = true;
+      }
+    };
+
+    await jsonld.expand(docWithUnMappedTerm, {expansionMap});
+
+    assert.equal(expansionMapCalled, true);
+  });
+
+  it('should be called on relative iri for id term', async () => {
+    const docWithRelativeIriId = {
+      '@context': {
+        'definedTerm': 'https://example.com#definedTerm'
+      },
+      '@id': "relativeiri",
+      definedTerm: "is defined"
+    };
+
+    let expansionMapCalled = false;
+    const expansionMap = info => {
+      if(info.relativeIri === 'relativeiri') {
+        expansionMapCalled = true;
+      }
+    };
+
+    await jsonld.expand(docWithRelativeIriId, {expansionMap});
+
+    assert.equal(expansionMapCalled, true);
+  });
+
+  it('should be called on relative iri for id term (nested)', async () => {
+    const docWithRelativeIriId = {
+      '@context': {
+        'definedTerm': 'https://example.com#definedTerm'
+      },
+      '@id': "urn:absoluteIri",
+      definedTerm: {
+        '@id': "relativeiri"
+      }
+    };
+
+    let expansionMapCalled = false;
+    const expansionMap = info => {
+      if(info.relativeIri === 'relativeiri') {
+        expansionMapCalled = true;
+      }
+    };
+
+    await jsonld.expand(docWithRelativeIriId, {expansionMap});
+
+    assert.equal(expansionMapCalled, true);
+  });
+
+  it('should be called on relative iri for aliased id term', async () => {
+    const docWithRelativeIriId = {
+      '@context': {
+        'id': '@id',
+        'definedTerm': 'https://example.com#definedTerm'
+      },
+      'id': "relativeiri",
+      definedTerm: "is defined"
+    };
+
+    let expansionMapCalled = false;
+    const expansionMap = info => {
+      if(info.relativeIri === 'relativeiri') {
+        expansionMapCalled = true;
+      }
+    };
+
+    await jsonld.expand(docWithRelativeIriId, {expansionMap});
+
+    assert.equal(expansionMapCalled, true);
+  });
+
+  it('should be called on relative iri for type term', async () => {
+    const docWithRelativeIriId = {
+      '@context': {
+        'definedTerm': 'https://example.com#definedTerm'
+      },
+      'id': "urn:absoluteiri",
+      '@type': "relativeiri",
+      definedTerm: "is defined"
+    };
+
+    let expansionMapCalled = false;
+    const expansionMap = info => {
+      if(info.relativeIri === 'relativeiri') {
+        expansionMapCalled = true;
+      }
+    };
+
+    await jsonld.expand(docWithRelativeIriId, {expansionMap});
+
+    assert.equal(expansionMapCalled, true);
+  });
```

Comment:
Trying to add (what I believe) is a missing test on typed literals via suggest mode here after this other test:

```suggestion
  it('should be called on relative iri for type term', async () => {
    const docWithRelativeIriId = {
      '@context': {
        'definedTerm': 'https://example.com#definedTerm'
      },
      'id': "urn:absoluteiri",
      '@type': "relativeiri",
      definedTerm: "is defined"
    };

    let expansionMapCalled = false;
    const expansionMap = info => {
      if(info.relativeIri === 'relativeiri') {
        expansionMapCalled = true;
      }
    };

    await jsonld.expand(docWithRelativeIriId, {expansionMap});

    assert.equal(expansionMapCalled, true);
  });
  
  it('should be called on relative iri for typed literal', async () => {
    const docWithRelativeIriId = {
      '@context': {
        'definedTerm': 'https://example.com#definedTerm'
      },
      'id': "urn:absoluteiri",
      definedTerm: {"@type": "relativeiri", "@value": "is defined"}
    };

    let expansionMapCalled = false;
    const expansionMap = info => {
      if(info.relativeIri === 'relativeiri') {
        expansionMapCalled = true;
      }
    };

    await jsonld.expand(docWithRelativeIriId, {expansionMap});

    assert.equal(expansionMapCalled, true);
  });  
```

Reply:
Interesting @dlongley when I run the test you added locally I get `jsonld.SyntaxError: Invalid JSON-LD syntax; an element containing "@value" and "@type" must have an absolute IRI for the value of "@type".` is there a clear delineation between when the expanded value of `@type` results in a relativeIri  is considered a syntax violation vs it not being?

Reply:
@tplooker,

I think the "clear delineation" is probably exactly the error message you quoted. You can't use a relative `@type` value in a typed literal value (vs. a node object). What we want to be testing there is what happens if a base URL is also passed in:

```js
await jsonld.expand(docWithRelativeIriId, {expansionMap, base: 'https://example.com'});
```

Hmm, this is also pointing to the fact that the current implementation in this PR isn't fully going to do what we want for signatures. We want the `expansionMap` to be called before a base is prepended -- and probably include the `base` value in the options passed to the handler function so an implementation can decide if it cares at that point.

We would certainly want jsonld-signatures lib's `expansionMap` to throw happen when expanding `@type` values (regardless of `base`'s value) -- because it is a common case that when people forget to define those in a document they are signing, they don't want any base to be applied to them giving them an unexpected (and wrong) definition.

Reply:
Thinking about this more...

While it would be easy enough to call the `expansionMap` prior to prepending base -- it will be tricky to get the property associated with the relative IRI value passed to the handler without a lot of changes (i.e., every call of `expandIri`) as we discussed above.

We could proceed with what we have for now -- and only throw when base isn't set -- which will cover *most* use cases. We recommend not setting `@base` in the JSON-LD signature world ... but now that I think about it, I recall @OR13 trying to use it somewhere. If someone uses `@base` *and* forgets to define an `@type` value, they could get burned because the URL will resolve to an absolute one (using their `@base` value) that may not be the one they intended.

So -- our options here are to make it so the `expansionMap` handler function gets the information it needs to decide. For example, we may want the logic in the jsonld-signatures `expansionMap` handler to be:

Is a relative `@type` value being expanded (prior to prepending base)? If so, throw an error. If not, see if a `base` is set. If one is set, don't throw, otherwise throw.

How do you think we should proceed?

Reply:
We could do something (maybe?) hacky and just pass whether an `@type` value is being expanded (a boolean value) to `expandIri` and default it to false. Then we'd only need to pass this where `@type` expansion occurs vs. having to update `expandIri` everywhere. That would cover the common use case -- and the only use case. All other URI values should be unrelated to vocabularies -- they'd just be purely node identifier values.

This approach may be the simplest way to fix it and cover most (all?) cases.

Reply:
Yeah, it would be good to expose whenever `@base` or `@vocab` are going to be applied. And I think my suggestion of saying whether or not it's an `@type` expansion should probably be sufficient (vs. having to pass the associated property for every value).

Reply:
See [here](https://github.com/digitalbazaar/jsonld.js/pull/452#issuecomment-845514850) for an update